### PR TITLE
TIG-2695 add withTransaction to CrudActor and replace startTxn/commitTxn in workloads

### PR DESCRIPTION
**Pull request by vrachev**
I'll probably do some DRY cleanup in crud actor but this should be functional (at least the yaml test works). Running the workloads in a patch now.

### CODE REVIEW

File: src/cast_core/src/CrudActor.cpp [hunk 1]
```diff
@@ -928,6 +930,85 @@ struct CommitTransactionOperation : public BaseOperation {
...
+        }
+        for (const auto& [k, txnOp] : opsInTxn) {
+            createTxnOps(txnOp, context, id);
+        }
+        if (opNode["Options"]) {
+            _options = opNode["Options"].to<mongocxx::options::transaction>();
+        }
+    }
+
+    void createTxnOps(const Node& txnOp, PhaseContext& context, ActorId id) {
```
Comment (aggrand) [inline src/cast_core/src/CrudActor.cpp:972]: Can this be in the private section?

File: src/cast_core/src/CrudActor.cpp [hunk 2]
```diff
@@ -928,6 +930,85 @@ struct CommitTransactionOperation : public BaseOperation {
...
+            : BaseOperation(context, opNode),
+              _onSession{onSession},
+              _collection{std::move(collection)},
+              _operation{operation} {
+        auto& opsInTxn = opNode["OperationsInTransaction"];
+        if (!opsInTxn.isSequence()) {
+            BOOST_THROW_EXCEPTION(InvalidConfigurationException(
+                                          "'WithTransaction' requires an 'OperationsInTransaction' node of sequence type."));
+        }
+        for (const auto& [k, txnOp] : opsInTxn) {
```
Comment (aggrand) [inline src/cast_core/src/CrudActor.cpp:964]: (nit) `const auto&&`, I think forward-referencing in ranged for loops is idiomatic in genny.

**Review by aggrand — CHANGES_REQUESTED** [review]
Great job tackling this! Mostly had a few nitpicks. Main question was regarding capturing the exception when the transaction fails.

Also, it might be a good idea to add a crud actor yaml test [here](https://github.com/mongodb/genny/blob/334f61ce6391ce464c0888180f474a5665095d7b/src/cast_core/test/CrudActorYamlTests.yml#L1). At least as a basic sanity check that the operations inside the transaction end up being performed.

If you want to locally end to end test the metrics, one approach is to use `curator ftdc` to export the metrics data and examine it. What I usually did for `InsertRemove` is do a single iteration, single threaded run then hand-check the data. Not sure if there'd be a good way to force the transaction to fail in order check for multiple "success" outputs for operations inside the transaction. Maybe temporarily adding a thread-sleep inside the callback, then in the mongo shell doing `rs.stepdown()`? (I presume that'll cancel the transaction?) Another approach is to compare the metrics data in patch builds from this PR with the same task on the current waterfall.

File: src/cast_core/src/CrudActor.cpp (hunk 1 above)
Comment (vrachev) [reply src/cast_core/src/CrudActor.cpp]: 👍

File: src/cast_core/src/CrudActor.cpp (hunk 2 above)
Comment (vrachev) [reply src/cast_core/src/CrudActor.cpp]: 👍

File: src/cast_core/src/CrudActor.cpp [hunk 3]
```diff
@@ -1028,6 +1110,21 @@ std::unordered_map<std::string, OpCallback&> opConstructors = {
     {"updateMany", baseCallback<BaseOperation, OpCallback, UpdateManyOperation>},
     {"replaceOne", baseCallback<BaseOperation, OpCallback, ReplaceOneOperation>}};
 
+std::unordered_map<std::string, OpCallback&> getOpConstructors(std::string mode) {
```
[the remark this answers is not in the document]
Comment (vrachev) [reply src/cast_core/src/CrudActor.cpp]: Good points. I looked into several options but I ended up just simplifying completely and writing the check for excluded ops in the `createTxnOps` function, and not putting any logic in the `getOpConstructors` function.

File: src/workloads/docs/CrudActorTransaction.yml [hunk 4]
```diff
@@ -22,20 +22,20 @@ Actors:
...
-        - WriteCommand: updateOne
-          Filter: {a: 1}
-          Update: {$set: {a: 5}}
-        Options:
-          Ordered: true
-        # The OnSession: true flag tells the driver to use a logical session. The actor will use the
-        # same session for the duration of a phase.
-        OnSession: true
-    - OperationName: commitTransaction
+        OperationsInTransaction:
```
[the remark this answers is not in the document]
Comment (vrachev) [reply src/workloads/docs/CrudActorTransaction.yml:27]: I added 

```        
ThrowOnFailure: false  # Transactions are allowed to fail.
RecordFailure: true  # We still want to record the metrics even if the transaction is rolled back.
```

above

**Review by vrachev — COMMENTED** [review]
Thanks for the review @aggrand! RFAL

> Also, it might be a good idea to add a crud actor yaml test here. At least as a basic sanity check that the operations inside the transaction end up being performed.

I did add a yaml test but you probably missed it in the midst of all the yaml files 🙂

File: src/cast_core/test/CrudActorYamlTests.yml [hunk 5]
```diff
@@ -572,3 +572,21 @@ Tests:
     OutcomeCounts:
       - Filter: {a: 1}
         Count: 1
+
```
Comment (aggrand) [inline src/cast_core/test/CrudActorYamlTests.yml:575]: Oof you're right I totally overlooked this as a workload yaml last time, my bad. This looks great!

**Review by aggrand — APPROVED** [review]
This LGTM assuming the metrics look right! Great work and clean code :)

**Comment by vrachev** [comment]
Thanks @aggrand, I checked the metrics with curator ftdc and they look correct

**Comment by vrachev** [comment]
evergreen merge

**Comment by vrachev** [comment]
evergreen merge

**Pull request merged**